### PR TITLE
Add decription on how you can customize image entrypoint

### LINKED ISSUE #14346: init scripts for dockerfile
**Description**

Add ability to process init scripts when starting airflow container (part of entrypoint script)

**Use case / motivation**

Similar to the [mysql entrypoint.sh](https://github.com/docker-library/mysql/blob/master/8.0/docker-entrypoint.sh#L369) allowing to run any number of init scripts before the scheduler/webserver is started. 
This can be usefull when using or extending the official docker image allowing to COPY scripts that can add connections, variables or any other airflow functionality automatically when starting the container

**Are you willing to submit a PR?**

yes. the PR should be relatively simple. As seen in the mysql example above

**Related Issues**

not the i found

### PR DESCRIPTION
**Pull request by potiuk**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: docs/docker-stack/entrypoint.rst [hunk 1]
```diff
@@ -194,6 +194,27 @@ If there are any other arguments - they are simply passed to the "airflow" comma
     optional arguments:
       -h, --help         show this help message and exit
 
+Add your own entrypoint
+-----------------------
+
+Airflow image executes a lot of steps in the entrypoint, but it does not do everything that you might
+want to run - and you might want to provide your own entrypoint which does any custom operations.
+Such ``custom`` entryopoint should be run after the main entrypoint. You can even execute
```
Comment (mik-laj) [inline docs/docker-stack/entrypoint.rst]: entryopoint=>entrypoint

File: docs/docker-stack/entrypoint.rst [hunk 2]
```diff
@@ -194,6 +194,27 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
+    FROM airflow::2.3.0.dev0
+    COPY my_entrypoint.sh /
+
+
+And then you can run this script by running the command:
+
+.. code-block:: bash
+
+  docker run -it apache/airflow:2.3.0.dev0-python3.6 bash -c "/my_entrypoint.sh"
+
```
Comment (mik-laj) [inline docs/docker-stack/entrypoint.rst:238]: Should we use exec?

Comment (potiuk) [reply docs/docker-stack/entrypoint.rst:238]: No need. We are using `dumb-init` which will propagate signals to all children by default (see the next chapter).

File: docs/docker-stack/entrypoint.rst (hunk 1 above)
Comment (uranusjr) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
Such *custom* entryopoint should be run after the main entrypoint. You can even execute
```

I'm assuming we don't actually mean to imply `custom` is some kind of code, so this switches to use more idiomatic emphasis format.

Comment (theodesp) [reply docs/docker-stack/entrypoint.rst]:
```suggestion
Such *custom* entrypoint should be run after the main entrypoint. You can even execute
```

File: docs/docker-stack/entrypoint.rst [hunk 3]
```diff
@@ -194,6 +194,27 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
       -h, --help         show this help message and exit
 
+Add your own entrypoint
+-----------------------
+
+Airflow image executes a lot of steps in the entrypoint, but it does not do everything that you might
+want to run - and you might want to provide your own entrypoint which does any custom operations.
+Such *custom* entrypoint should be run after the main entrypoint. You can even execute
+the usual components of airflow - scheduler, webserver, worker, triggerer in such script.
+The entrypoint can be addedd to the image by extending it:
```
Comment (potiuk) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
The entrypoint can be added to the image by extending it:
```

**Comment by yiga2** [comment]
I still believe there is an use case for executing stuff BEFORE the entrypoint. In our case, we override the DB pointer dynamically, depending on some values in the ENV -> can only be set at runtime. Hence must run **before** the entrypoint as the latter (`entrypoint_prod`) performs a DB check, which we want (+ create user).

Our way around it: build an image with our `my_entrypoint.sh` as entrypoint, which will do all ENV settings, and then exec the base entrypoint passing on any argument.

**Comment by potiuk** [comment]
Cool. Now I need just an approve from another committer :). Anyone?

File: docs/docker-stack/entrypoint.rst [hunk 4]
```diff
@@ -194,6 +194,69 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
+
+.. code-block:: bash
+
+  docker run -it apache/airflow:2.3.0.dev0-python3.6 bash -c "/my_entrypoint_extension.sh"
+
+Execute custom code before the Airflow entrypoint
+-------------------------------------------------
+
+If you want to execute some custom code before Airflow's entrypoint you can also add extra code executed
+before the Airflow entrypoint by replacing the entrypoint of Airflow and calling Airflow's entrypoint as
```
Comment (potiuk) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
before the Airflow one by using your custom script and calling Airflow's entrypoint as
```

File: docs/docker-stack/entrypoint.rst [hunk 5]
```diff
@@ -194,6 +194,69 @@ If there are any other arguments - they are simply passed to the "airflow" comma
     optional arguments:
       -h, --help         show this help message and exit
 
+Add your own entrypoint extension
+---------------------------------
+
+Airflow image executes a lot of steps in the entrypoint, but it does not do everything that you might
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
The Airflow image executes a lot of steps in the entrypoint, but it does not do everything that you might
```

File: docs/docker-stack/entrypoint.rst [hunk 6]
```diff
@@ -194,6 +194,69 @@ If there are any other arguments - they are simply passed to the "airflow" comma
     optional arguments:
       -h, --help         show this help message and exit
 
+Add your own entrypoint extension
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]: I'm not sure I'm following on this extension section. This is simply just COPYing an extra script into the image, the entrypoint wont actually run it, right? I'm not sure it's worth having this example if that's the case, and I think using "extension" is slightly misleading too.

Comment (potiuk) [reply docs/docker-stack/entrypoint.rst]: I think you are right, it's more about adding custom behaviour. I will reword it

Comment (potiuk) [reply docs/docker-stack/entrypoint.rst]: I reworded it, and put it as "second" option - explaining that this customizing the image behaviour.

File: docs/docker-stack/entrypoint.rst [hunk 7]
```diff
@@ -194,6 +194,72 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
+execution (A bit useless example but should give the reader an example of how you could use it).
+
+.. code-block:: bash
+
+    #!/bin/bash
+    export CONNECTION_CHECK_MAX_COUNT=${1}
+    shift
+    exec /entrypoint "${@}"
+
+Make sure, Airflow's entrypoint is run as ``exec /entrypoint "${@}"`` as the last command in your
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
Make sure Airflow's entrypoint is run with ``exec /entrypoint "${@}"`` as the last command in your
```

File: docs/docker-stack/entrypoint.rst [hunk 8]
```diff
@@ -194,6 +194,72 @@ If there are any other arguments - they are simply passed to the "airflow" comma
     optional arguments:
       -h, --help         show this help message and exit
 
+Execute custom code before the Airflow entrypoint
+-------------------------------------------------
+
+If you want to execute some custom code before Airflow's entrypoint you can also add extra code executed
+before the Airflow one by using your custom script and calling Airflow's entrypoint as
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
If you want to execute some custom code before Airflow's entrypoint you can by using
a custom script and calling Airflow's entrypoint as the
```

File: docs/docker-stack/entrypoint.rst [hunk 9]
```diff
@@ -194,6 +194,72 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
+propagation (See the next chapter).
+
+
+.. code-block:: Dockerfile
+
+    FROM airflow::2.3.0.dev0
+    COPY my_entrypoint.sh /
+    ENTRYPOINT ["/usr/bin/dumb-init", "--", "/my_entrypoint.sh"]
+
+Your entrypoint might for example modify or add variables on the flight. For example the below
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
Your entrypoint might for example modify or add variables on the fly. For example the below
```

File: docs/docker-stack/entrypoint.rst [hunk 10]
```diff
@@ -194,6 +194,72 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
+
+Make sure, Airflow's entrypoint is run as ``exec /entrypoint "${@}"`` as the last command in your
+custom entrypoint. This way signals will be properly propagated and arguments will be passed
+to the entrypoint as usual (you can use ``shift`` as above if you need to pass some extra
+arguments. Note that passing secret values this way or storing secrets inside the image is a bad
+idea from security point of view - as both image and parameters to run the image with are accessible
+to anyone who has access to logs of your Kubernetes or image registry.
+
+Also be aware that code executed before Airflow's entrypoint should not create any files or
+directories inside the image and not everything might work the same way as after it is executed.
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
directories inside the container and everything might not work the same way when it is executed.
```

File: docs/docker-stack/entrypoint.rst [hunk 11]
```diff
@@ -194,6 +194,72 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
+Before Airflow entrypoint is executed, the following functionalities are not available:
+
+* umask is not set properly to allow ``group`` write access
+* user is not yet created in /etc/passwd in case arbitrary user is used to run the image
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
* user is not yet created in ``/etc/passwd`` if an arbitrary user is used to run the image
```

File: docs/docker-stack/entrypoint.rst [hunk 12]
```diff
@@ -194,6 +194,72 @@ If there are any other arguments - they are simply passed to the "airflow" comma
...
+* the database and brokers might not be available yet
+
+Adding custom image behaviour
+-----------------------------
+
+The Airflow image executes a lot of steps in the entrypoint, and sets the right environment, but
+you might want to run additional code after the entrypoint creating the user, set the umask, setting the
+variables and checking that database is runnint.
```
Comment (jedcunningham) [inline docs/docker-stack/entrypoint.rst]:
```suggestion
you might want to run additional code after the entrypoint creates the user, sets the umask, sets
variables and checks that database is running.
```

**Review by jedcunningham — APPROVED** [review]
Overall looks good, it makes sense as it is written now 👍

**Pull request merged**